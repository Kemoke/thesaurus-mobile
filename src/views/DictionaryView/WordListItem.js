import React from 'react';
import {FlatList, View} from "react-native";
import {ListItem} from "react-native-elements";

type WordListItemProps = {
    item: {
        word: string,
        synonyms: Array<string>
    }
}

export default class WordListItem extends React.Component<WordListItemProps>{
    state = {
        expanded: false
    };

    renderSynonymItem = ({item}) => (
        <View>
            <ListItem containerStyle={{backgroundColor: '#EEEEEE'}}
                      contentContainerStyle={{paddingLeft: 16}}
                      title={item}
                      bottomDivider/>
        </View>
    );

    toggleSynonymList = () => this.setState({expanded: !this.state.expanded});

    keyExtractor = (item) => item;

    render() {
        const {item} = this.props;
        const {expanded} = this.state;
        return (
            <View>
                <ListItem title={item.word}
                          badge={{value: item.synonyms.length}}
                          bottomDivider
                          chevron
                          testID='wordItem'
                          onPress={this.toggleSynonymList}/>
                {expanded && <FlatList keyExtractor={this.keyExtractor}
                                       data={item.synonyms}
                                       renderItem={this.renderSynonymItem}/>}
            </View>
        );
    }
}