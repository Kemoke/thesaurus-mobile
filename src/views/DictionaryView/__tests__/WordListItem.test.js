import React from 'react';
import {shallow} from "enzyme";
import WordListItem from "../WordListItem";

describe('Word List Item', function () {
    it('should render properly', function () {
        const item = {
            word: "test",
            synonyms: ["test1", "test2"]
        };
        const component = shallow(
            <WordListItem item={item}/>
        );
        expect(component.instance().keyExtractor("test1")).toEqual("test1");
        expect(component).toMatchSnapshot();
    });
    it('should render expanded properly', function () {
        const item = {
            word: "test",
            synonyms: ["test1", "test2"]
        };
        const component = shallow(
            <WordListItem item={item}/>
        );
        component.setState({expanded: true});
        expect(component).toMatchSnapshot();
    });
    it('should handle toggling list expansion', function () {
        const item = {
            word: "test",
            synonyms: ["test1", "test2"]
        };
        const component = shallow(
            <WordListItem item={item}/>
        );
        component.findWhere(node => node.prop('testID') === "wordItem").props().onPress();
        expect(component.state("expanded")).toEqual(true);
        component.findWhere(node => node.prop('testID') === "wordItem").props().onPress();
        expect(component.state("expanded")).toEqual(false);
    });
    it('should render synonym item properly', function () {
        const item = {
            word: "test",
            synonyms: ["test1", "test2"]
        };
        const component = shallow(
            <WordListItem item={item}/>
        );
        const synonymItem = shallow(component.instance().renderSynonymItem({item: "test1"}));
        expect(synonymItem).toMatchSnapshot();
    });
});