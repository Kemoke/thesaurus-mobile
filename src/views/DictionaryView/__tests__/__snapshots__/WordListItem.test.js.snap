// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word List Item should render expanded properly 1`] = `
<Component>
  <Themed.ListItem
    badge={
      Object {
        "value": 2,
      }
    }
    bottomDivider={true}
    chevron={true}
    onPress={[Function]}
    testID="wordItem"
    title="test"
  />
  <FlatList
    data={
      Array [
        "test1",
        "test2",
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Word List Item should render properly 1`] = `
<Component>
  <Themed.ListItem
    badge={
      Object {
        "value": 2,
      }
    }
    bottomDivider={true}
    chevron={true}
    onPress={[Function]}
    testID="wordItem"
    title="test"
  />
</Component>
`;

exports[`Word List Item should render synonym item properly 1`] = `
<View>
  <Themed.ListItem
    bottomDivider={true}
    containerStyle={
      Object {
        "backgroundColor": "#EEEEEE",
      }
    }
    contentContainerStyle={
      Object {
        "paddingLeft": 16,
      }
    }
    title="test1"
  />
</View>
`;
